presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -0,0 +1,148 @@
+{
+    public PlanNode optimize(PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan, new Context());
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<MergeIdenticalWindows.Context>
+    {
+        @Override
+        public PlanNode visitWindow(
+                WindowNode node,
+                RewriteContext<MergeIdenticalWindows.Context> context)
+        {
+            Context mergeContext = context.get();
+            boolean retainNode = mergeContext.collectFunctions(node);
+
+            // Don't merge window functions across nodes that aren't WindowNodes
+            if (node.getClass().isInstance(node.getSource())) {
+                node = (WindowNode) context.defaultRewrite(node, mergeContext);
+            }
+            else {
+                node = (WindowNode) context.defaultRewrite(node, new Context());
+            }
+
+            // WindowFunctions contained in this node will be merged in some node further up the tree.
+            if (!retainNode) {
+                return node.getSource();
+            }
+
+            Context.FunctionPair collectedFunctions = mergeContext.getCollectedFunctions(node);
+
+            return new WindowNode(
+                    node.getId(),
+                    node.getSource(),
+                    node.getSpecification(),
+                    collectedFunctions.windowFunctions.build(),
+                    collectedFunctions.signatures.build(),
+                    node.getHashSymbol(),
+                    node.getPrePartitionedInputs(),
+                    node.getPreSortedOrderPrefix());
+        }
+    }
+
+    private static class Context
+    {
+        private Map<WindowNode.Specification, FunctionPair> pairMap = new HashMap<>();
+
+        private boolean collectFunctions(WindowNode window)
+        {
+            WindowNode.Specification key = window.getSpecification();
+            FunctionPair pair = pairMap.get(key);
+
+            if (pair == null) {
+                pair = new FunctionPair();
+            }
+
+            pair.windowFunctions.putAll(window.getWindowFunctions());
+            pair.signatures.putAll(window.getSignatures());
+            return pairMap.put(key, pair) == null;
+        }
+
+        private FunctionPair getCollectedFunctions(WindowNode window)
+        {
+            WindowNode.Specification key = window.getSpecification();
+            checkState(
+                    pairMap.containsKey(key),
+                    "No pair for window specification. Tried to merge the same window twice?");
+            return pairMap.get(key);
+        }
+
+        private static class FunctionPair
+        {
+            final ImmutableMap.Builder<Symbol, FunctionCall> windowFunctions = new ImmutableMap.Builder<>();
```

Comment:
I'm not generally a fan of the direct access to class internals approach. I would still prefer this to have getters instead and make the fields private final.